### ISSUE TITLE
`rustc` doesn't produce warning on unused `mut` if value is taken by reference.

### ISSUE THREAD

**Issue by popzxc** [rust-lang/rust#66687]
Currently, if function takes a mutable argument, but the mutability is not required by the function body, `rustc` procudes a warning: "warning: variable does not need to be mutable".

However, if argument is taken by a mutable reference, the warning is not produced.

Here's example on the playground: [link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=1b454d8566039c0813ff83b91420330b).

Code from the playground:

```rust
struct Example;

impl Example {
    fn get_val(&self) -> u8 {
        42
    }
}

fn immut_ref(val: &Example) -> u8 {
    val.get_val()
}

// Will not spawn warning.
fn mut_ref(val: &mut Example) -> u8 {
    val.get_val()
}

// Will spawn warning (warning: variable does not need to be mutable).
fn mut_val(mut val: Example) -> u8 {
    val.get_val()
}

fn main() {
    let mut example = Example;
    
    immut_ref(&example);
    
    mut_ref(&mut example);
    
    mut_val(example);
}
```

**Comment by jonas-schievink** [comment]
The lint only warns for unnecessary `mut` bindings, not mutability in the type of any variables. Clippy has a lint for this though: https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_mut_passed

As such, closing.

**Comment by popzxc** [comment]
The linked lint is not really related: it warns about `&mut` reference passed to the place where immutable reference is enough. In the code I've pointed, it's in the signature of the function (and not required).

**Comment by jonas-schievink** [comment]
Ah, you're right. I thought clippy also had a lint for that, but it doesn't look like it. It still seems like Clippy is a good place to prototype this sort of lint.

**Comment by popzxc** [comment]
Sure.
